architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,580 @@
+
+*Note: the exact set of tooling implementations are out of scope for this document.*
+
+## User Experience (if applicable)
+
+See API design. Tooling and user-experiences are out of scope for this document.
+
+## Design
+
+### High Level Design
+
+The resource type registration APIs allow a user to control the set of resource types known to UCP. Since UCP is responsible for routing, it needs to know the set of resource providers and resource types so that it can send them traffic. 
+
+#### Resource Types
+
+The resource type registration APIs add two new resource types:
+
+- `System.Resources/resourceProviders` for resource providers.
+- `System.Resources/resourceProviders/resourceTypes` for resource types.
+
+These new resource types define the full CRUDL set of APIs. These APIs are optimized for operations that need to modify or deeply understand resource types. 
+
+`resourceTypes` is our first nested resource type in UCP/Radius. This is a good usecase for a nested resource types because:
+
+- The `resourceTypes` body will be be large due to the inclusion of an OpenAPI schema. 
+- Users need granular registration and unregistration of individual resource types. (See scenario 2 above).
+
+---
+
+These new resource types are scoped to the Radius plane. 
+
+> eg: `/planes/radius/local/providers/System.Resources/resourceProviders`
+
+The rationale for this is to support registration/unregistration per-tenant of Radius rather than per-installation.
+
+#### Resource Type Queries
+
+The proposal also defines two new **non-resource** APIs:
+
+- `GET /planes/radius/{plane}/providers/`: list all resource providers.
+- `GET /planes/radius/{plane}/providers/{provider}` shows a specific resource provider.
+
+These APIs are optimized for queries and are consistent with similar APIs in ARM. To understand why this is important, remember that the OpenAPI schema for a resource type can be quite large. Some functionality in Radius will need to query resource types frequently, consider the need for the `rad recipe register` to validate a resource type. Building a separate API for queries allows us to optimize the behavior.
+
+These APIs can also be implemented consistently across all planes in the future. For example Azure doesn't support user-defined types, but it does support looking up the set of resource types.
+
+Lastly, there will still be some system-defined types. For example `System.Resources/resourceProviders` is a system defined type. The query API can return these, but the `System.Resources/resourceProviders` cannot. Imagine for a minute what would happen if users could remove the capability to specify a type.
+
+**Open Question:** Is *query* the right term for this? The term will show up in docs and in the TypeSpec + generated client + our code. Let's pick the term that's least confusing to us. I'm open to suggestions but I'll just roll with *query* unless there's a better idea.
```

Comment:
The name makes sense to me, but the user will be able to do **Command** and **Query**, right? By **Command** I mean adding a new Resource Type. Why not call it Resource Type API or Custom Resource Type API?

Reply:
Both of these terms are pretty ambiguous IMO. 

- Resource provider query API - what I'm proposing. 
- Resource Type API - doesn't feel distinct from `System.Resources/resourceProviders/resourceTypes`.
- Custom Resource Type - I don't think the word custom helps us tell these two APIs apart.

Reply:
I changed this to `resource provider summary`. Now that the other APIs are more complex, it makes sense to call it a summary. It's not **all** of the info, just the most commonly-used parts.